Feature pass MPI comm from Python script to AMReX during initialization

KZhu-ME opened the pull request:
Added ability to pass MPI communicator from input Python script to AMReX when initializing WarpX. Passing the communicator is optional, and original functionality remains the same if no communicator is specified to be passed from the Python script.

dpgrote commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -20,6 +20,10 @@
     # --- before libwarpx is loaded (though don't know why)
     from mpi4py import MPI
+    if MPI.sizeof(MPI.Comm) == ctypes.sizeof(ctypes.c_int):
+        MPI_Comm = ctypes.c_int
```
Can the name be changed from `MPI_Comm` to `_MPI_Comm_type` for clarity? Also, it will need to be defined if mpi4py cannot be imported (in the `except` clause).

KZhu-ME replied:
Yep! I will get it fixed and pushed by tomorrow.

dpgrote reviewed (commented):
Thanks for this PR! This would be a nice feature to have. I have only the one minor comment.

ax3l commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -20,8 +20,12 @@
     # --- If mpi4py is going to be used, this needs to be imported
     # --- before libwarpx is loaded (though don't know why)
```
>     # --- If mpi4py is going to be used, this needs to be imported
>     # --- before libwarpx is loaded (though don't know why)

In case we want to update the comment:
That's because importing `mpi4py` calls `MPI_Init`.

ax3l commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -234,17 +241,22 @@ def get_nattr():
+    else:
+        comm_ptr = MPI._addressof(mpi_comm)
+        comm_val = _MPI_Comm_type.from_address(comm_ptr)
+        libwarpx.amrex_init_with_inited_mpi(argc, argvC, comm_val)
```
Runtime issue if build with `-DWarpX_MPI=OFF`:
> AttributeError: /home/runner/.local/lib/python3.8/site-packages/pywarpx/libwarpx.3d.so: undefined symbol: amrex_init_with_inited_mpi

We should guard this section so that we do not call missing symbols in serial builds.

ax3l reviewed (commented):
Thanks, great idea!

Can you also add a small test that uses this feature?

dpgrote commented on Source/Python/WarpXWrappers.cpp:
```diff
@@ -130,6 +130,13 @@ extern "C"
     {
         warpx_amrex_init(argc, argv, true, mpicomm);
     }
+#else
```
I think this can be simplified by removing the `#ifdef BL_USE_MPI` macro (and the `else` block as well) so that the routine is always compiled. As you say, this routine shouldn't be called in serial but it needs to be declared, so a special version of it is not needed for serial.

dpgrote commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -18,10 +18,14 @@
+    else:
+        _MPI_Comm_type = ctypes.c_void_p
 except ImportError:
-    pass
```
Perhaps add this `MPI = None`, so it can be used as a check.

dpgrote commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -234,17 +241,22 @@ def get_nattr():
         argvC[i] = ctypes.create_string_buffer(enc_arg)
 
-    libwarpx.amrex_init(argc, argvC)
+    if mpi_comm is None:
```
Perhaps do this `if mpi_comm is None or MPI is None:` to add an extra check.

kzhu-ME commented:
Thanks for the input! Changes have been implemented that should fix the compile errors. I'm currently working on writing a test for this feature.

ax3l commented on Source/Python/WarpXWrappers.h:
```diff
@@ -37,9 +37,9 @@ extern "C" {
     void amrex_init (int argc, char* argv[]);
 
-#ifdef BL_USE_MPI
+    #ifdef BL_USE_MPI
```
I know it's old code, but the new variable is called:
```suggestion
#ifdef AMREX_USE_MPI
```

these days :)

ax3l commented on Source/Python/WarpXWrappers.h:
```diff
@@ -37,9 +37,9 @@ extern "C" {
+    #ifdef BL_USE_MPI
     void amrex_init_with_inited_mpi (int argc, char* argv[], MPI_Comm mpicomm);
-#endif
+    #endif
```
```suggestion
#endif
```

ax3l commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -18,13 +18,15 @@
         _MPI_Comm_type = ctypes.c_int
     else:
         _MPI_Comm_type = ctypes.c_void_p
+        MPI = None
```
I am not sure this line is correct.
We should maybe add comments here: the first branch is MPICH-like flavors, the else is OpenMP-like flavors (or vice versa, need to check :) ).

kzhu-ME commented:
I've implemented all the suggested changes and added a test for the feature. Due to the nature of the feature, the assert statements are in the input PICMI file for the test, as I needed to check during the simulation how many processors and which processors were being used. This means that if the test fails, it says that the test crashes instead of fails.

kzhu-ME commented:
All requested changes have been implemented, and this PR is ready for review.

dpgrote reviewed (commented):
Looking at this some more, I think more changes could be added, particularly things that will simplify the example python input file. The `mpi_comm` argument should be added to the `Simulations.step` method in `picmi.py` and also to the intermediate routine so it gets passed to the call to `pywarpx.warpx.init`. The input file would then not need to call `initialize_inputs` and `initialize_warpx`.

The python test should not import _libwarpx directly but should access its methods through the wx module, doing `from pywarpx import wx`, and this should only be done at the end of the input file just before it calls `getNProcs`. With this structure, then importing geometry is not needed as well as the import at line 34.

You can also comment out the line call `write_input_file` since that is not needed.

To make the test case more interesting, maybe have it run two difference cases simultaneously on the two processors, changing a parameter, perhaps the plasma density. The two simulations can write to two different diags files, `diag1` and `diag2` and the analysis script can check that both simulations ran and get the appropriately different results.

dpgrote commented on Examples/Tests/pass_mpi_communicator/PICMI_inputs_2d.py:
```diff
@@ -0,0 +1,254 @@
+from mpi4py import MPI
+from pywarpx import picmi
+constants = picmi.constants
+# from pywarpx import geometry
```
These lines should be removed.

dpgrote commented on Examples/Tests/pass_mpi_communicator/PICMI_inputs_2d.py:
```diff
@@ -0,0 +1,254 @@
+    color = 1
+new_comm = comm_world.Split(color)
+
+if color == 0:
```
Instead of duplicating all of the setup, could you do something like
`plasma_density = [1.e18, 1.e19][color]`
and
`diagname = f'diag{color+1}'`
and have both processors run the same code? Otherwise, this looks good.

KZhu-ME replied:
I'll fix that. Yeah I've tested with both processors running the same code, and they get the same outputs as expected.

KZhu-ME commented:
@dpgrote I implemented all the changes, and this PR is ready for review when you have time.

ax3l commented on Examples/Tests/pass_mpi_communicator/analysis.py:
```diff
@@ -0,0 +1,79 @@
+import sys
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
```
That's the unused import that CI marked :)
```suggestion
```

ax3l commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -18,10 +18,15 @@
-    # --- before libwarpx is loaded (though don't know why)
+    # --- before libwarpx is loaded, because mpi4py calls MPI_Init
     from mpi4py import MPI
+    if MPI._sizeof(MPI.Comm) == ctypes.sizeof(ctypes.c_int):
+        _MPI_Comm_type = ctypes.c_int
+    else:
+        _MPI_Comm_type = ctypes.c_void_p
```
Is it ok if we add an inline code comment here?
If I see the code correctly, you are switching essentially between OpenMPI and MPICH flavors based on the type of the communicator? Is that correct? Let's add a few lines here to conserve that knowledge for future readers? :)

KZhu-ME commented:
Thanks for the comments! The changes have been pushed.

ax3l reviewed (approved):
Thanks a lot, looks great :+1:

ax3l commented:
@dpgrote I leave this open for you for the final look/merge.

dpgrote reviewed (approved):
Looks good to me! Thanks for the PR and making the recommended fixes.
